PROTOCOL.grl

```diff
@@ -5,22 +5,56 @@
 This file allows to selectively revoke one or several grants from a
 certificate without revoking all the grants or the certificate itself.
 
+Note: Because of the added flexibility, the GRL is not as efficient at storing
+list of certificates for which all grants are revoked. This case is better
+handled by revoking the certificate itself in an OpenSSH KRL.
+
+
 ## Format
 
-The format of this file follows the openssh KRL format for certificates
-(KRL_SECTION_CERTIFICATES) by serial + bitmap (KRL_SECTION_CERT_SERIAL_BITMAP).
+The hiba-chk is a short lived helper invoked by OpenSSH and will need to parse
+the GRL file for every invocation. Therefore, the format has been optimized for
+fast SSH certificate serial lookups.
+
+The GRL starts with a header followed by two sections: serials list and grant
+revocation maps. Each section starts with a uint64 that store the section size.
+
+The file also contains a trailer magic.
+
+```
+	uint32 magic
+	uint32 version
+	uint32 min_version
+	uint64 timestamp
+	string comment
+```
+
+Where magic is defined as:
+
+```
+#define HIBA_GRL_MAGIC	0x4847524c
+```
+
+The next section is contains a sorted list of certificate serial numbers
+associated with and index. Each entry has a fixed size in order to allow quick
+lookups. The index is used to find the corresponding bitmap of revoked grants
+provided in the next section.
+
+```
+	uint64 serial
+	uint64 offset
+```
 
-In this case, while the format is consistent with openssh KRL file, the
-semantic is slightly different. The GRL serial points to a given certificate,
-while the bitmap is used to target the grants inside this certificate.
+The last section contains a list of bitmaps. Each bit of the map represents a
+grant from the certificate. This means that the order of grants in a
+certificate matters.
 
 ```
-	uint64  serial_offset
-	mpint   revoked_grants_bitmap
+	mpint  bitmap
```

Comment:
I believe the trailer magic is missing here.

Reply:
Done